### PR TITLE
Adding new RDK plugin for minidump collecting purposes

### PR DESCRIPTION
**Pull request by g3t0r1x**
### Description
XIONE-4968: Minidump core collection for WPEWebkit
Reason for change: Enabling minidump collection for dobby container apps

### Test Procedure
Test if sending linux signal with core dump for WPEWebkit
    based apps generates and uploads minidump file to backtrace portal

### Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (doesn't fit into the above categories - e.g. documentation updates)

### Requires Bitbake Recipe changes?
- [ ] The base Bitbake recipe (`meta-rdk-ext/recipes-containers/dobby/dobby.bb`) must be modified to support the changes in this PR (beyond updating `SRC_REV`)
- [x] Requires change https://gerrit.teamccp.com/#/c/539640/ to be merged before
- [x] Requires change https://gerrit.teamccp.com/#/c/539639/ to be merged before

### CODE REVIEW

File: rdkPlugins/Minidump/source/Minidump.cpp
Comment (TeknoVenus) [inline rdkPlugins/Minidump/source/Minidump.cpp:227]: Use `!` please

**Review by TeknoVenus — CHANGES_REQUESTED** [review]

**Comment by TeknoVenus** [comment]
@g3t0r1x Please fix failing build and add CMake flag to enable/disable the plugin (it should be disabled by default please).

Please also update the `.lgtm.yml` and `.github/workflows/build.yml` to enable the plugin during the automated build/analysis.

**Comment by TeknoVenus** [comment]
As discussed, I've identified some issues:

* If you enable the minidump on a container where the process is **not** compiled with breakpad support (and hence won't generate a minidump file), the plugin creates a blank 11M file in `/opt/minidumps/minidump.dmp`

  * Create container that runs `sleep 60` with minidump plugin enabled

  * Run `kill -11 <sleep pid>`

  * Plugin will report `Minidump copied to: /opt/minidumps/minidump.dmp`

  * File is 11M of nothing:

    ```
    vagrant@dobby-vagrant:/opt/minidumps$ ls -alh
    total 11M
    drwxrwxr-- 2 vagrant vagrant 4.0K Jul 13 12:45 .
    drwxr-xr-x 8 root    root    4.0K Jul 13 12:35 ..
    -rw-r--r-- 1 vagrant vagrant  11M Jul 13 12:45 minidump.dmp
    
    vagrant@dobby-vagrant:/opt/minidumps$ hexdump ./minidump.dmp 
    0000000 0000 0000 0000 0000 0000 0000 0000 0000
    *
    0a00001
    ```

* A new error is logged by DobbyDaemon during container shutdown (both during graceful shutdown and after SIGSEGV)

  ```
  0000001337.042908 ERR: < M:DobbyStream.cpp F:~DobbyBufferStream L:157 > failed to close memfd (9 - Bad file descriptor)
  ```

* When container is shut down gracefully with `DobbyTool stop`, minidump plugin prints an error:

  ```
  0000001420.326654 NFO: < M:DobbyRdkPluginManager.cpp F:runPlugins L:716 > Running minidump plugin
  0000001420.337533 ERR: < M:AnonymousFile.cpp F:copyContentTo L:161 > Cannot open /opt/minidumps/test.dmp
  0000001420.339195 ERR: < M:DobbyRdkPluginManager.cpp F:runPlugins L:724 > Required plugin minidump postHalt hook has failed
  ```

* Invalid write errors detected by Valgrind

```
0000000663.899468 NFO: < M:DobbyRdkPluginManager.cpp F:runPlugins L:716 > Running minidump plugin
==14000== Thread 8 AI_SIGMONITOR:
==14000== Invalid write of size 1
==14000==    at 0xC038C03: AnonymousFile::copyContentTo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (AnonymousFile.cpp:147)
==14000==    by 0xC036D6C: Minidump::postHalt() (Minidump.cpp:130)
==14000==    by 0x64B7AC: DobbyRdkPluginManager::executeHook(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, IDobbyRdkPlugin::HintFlags) const (DobbyRdkPluginManager.cpp:611)
==14000==    by 0x64BE6E: DobbyRdkPluginManager::runPlugins(IDobbyRdkPlugin::HintFlags const&) const (DobbyRdkPluginManager.cpp:717)
==14000==    by 0x534271: DobbyManager::onPostHaltHook(std::unique_ptr<DobbyContainer, std::default_delete<DobbyContainer> > const&) (DobbyManager.cpp:1830)
==14000==    by 0x534E99: DobbyManager::onChildExit() (DobbyManager.cpp:2142)
==14000==    by 0x535789: DobbyManager::runcMonitorThread() (DobbyManager.cpp:2338)
==14000==    by 0x54747A: void std::__invoke_impl<void, void (DobbyManager::* const&)(), DobbyManager*>(std::__invoke_memfun_deref, void (DobbyManager::* const&)(), DobbyManager*&&) (functional:227)
==14000==    by 0x5473F3: std::result_of<void (DobbyManager::* const&(DobbyManager*&&))()>::type std::__invoke<void (DobbyManager::* const&)(), DobbyManager*>(void (DobbyManager::* const&)(), DobbyManager*&&) (functional:251)
==14000==    by 0x547395: decltype (__invoke((*this)._M_pmf, (forward<DobbyManager*>)({parm#1}))) std::_Mem_fn_base<void (DobbyManager::*)(), true>::operator()<DobbyManager*>(DobbyManager*&&) const (functional:604)
==14000==    by 0x547360: void std::_Bind_simple<std::_Mem_fn<void (DobbyManager::*)()> (DobbyManager*)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) (functional:1391)
==14000==    by 0x54706F: std::_Bind_simple<std::_Mem_fn<void (DobbyManager::*)()> (DobbyManager*)>::operator()() (functional:1380)
==14000==  Address 0xe341041 is 0 bytes after a block of size 10,485,761 alloc'd
==14000==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14000==    by 0xC038B36: AnonymousFile::copyContentTo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (AnonymousFile.cpp:139)
==14000==    by 0xC036D6C: Minidump::postHalt() (Minidump.cpp:130)
==14000==    by 0x64B7AC: DobbyRdkPluginManager::executeHook(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, IDobbyRdkPlugin::HintFlags) const (DobbyRdkPluginManager.cpp:611)
==14000==    by 0x64BE6E: DobbyRdkPluginManager::runPlugins(IDobbyRdkPlugin::HintFlags const&) const (DobbyRdkPluginManager.cpp:717)
==14000==    by 0x534271: DobbyManager::onPostHaltHook(std::unique_ptr<DobbyContainer, std::default_delete<DobbyContainer> > const&) (DobbyManager.cpp:1830)
==14000==    by 0x534E99: DobbyManager::onChildExit() (DobbyManager.cpp:2142)
==14000==    by 0x535789: DobbyManager::runcMonitorThread() (DobbyManager.cpp:2338)
==14000==    by 0x54747A: void std::__invoke_impl<void, void (DobbyManager::* const&)(), DobbyManager*>(std::__invoke_memfun_deref, void (DobbyManager::* const&)(), DobbyManager*&&) (functional:227)
==14000==    by 0x5473F3: std::result_of<void (DobbyManager::* const&(DobbyManager*&&))()>::type std::__invoke<void (DobbyManager::* const&)(), DobbyManager*>(void (DobbyManager::* const&)(), DobbyManager*&&) (functional:251)
==14000==    by 0x547395: decltype (__invoke((*this)._M_pmf, (forward<DobbyManager*>)({parm#1}))) std::_Mem_fn_base<void (DobbyManager::*)(), true>::operator()<DobbyManager*>(DobbyManager*&&) const (functional:604)
==14000==    by 0x547360: void std::_Bind_simple<std::_Mem_fn<void (DobbyManager::*)()> (DobbyManager*)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) (functional:1391)
==14000== 
==14000== Syscall param write(buf) points to uninitialised byte(s)
==14000==    at 0x597F4BD: ??? (syscall-template.S:84)
==14000==    by 0xC038DD0: AnonymousFile::copyContentTo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (AnonymousFile.cpp:167)
==14000==    by 0xC036D6C: Minidump::postHalt() (Minidump.cpp:130)
==14000==    by 0x64B7AC: DobbyRdkPluginManager::executeHook(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, IDobbyRdkPlugin::HintFlags) const (DobbyRdkPluginManager.cpp:611)
==14000==    by 0x64BE6E: DobbyRdkPluginManager::runPlugins(IDobbyRdkPlugin::HintFlags const&) const (DobbyRdkPluginManager.cpp:717)
==14000==    by 0x534271: DobbyManager::onPostHaltHook(std::unique_ptr<DobbyContainer, std::default_delete<DobbyContainer> > const&) (DobbyManager.cpp:1830)
==14000==    by 0x534E99: DobbyManager::onChildExit() (DobbyManager.cpp:2142)
==14000==    by 0x535789: DobbyManager::runcMonitorThread() (DobbyManager.cpp:2338)
==14000==    by 0x54747A: void std::__invoke_impl<void, void (DobbyManager::* const&)(), DobbyManager*>(std::__invoke_memfun_deref, void (DobbyManager::* const&)(), DobbyManager*&&) (functional:227)
==14000==    by 0x5473F3: std::result_of<void (DobbyManager::* const&(DobbyManager*&&))()>::type std::__invoke<void (DobbyManager::* const&)(), DobbyManager*>(void (DobbyManager::* const&)(), DobbyManager*&&) (functional:251)
==14000==    by 0x547395: decltype (__invoke((*this)._M_pmf, (forward<DobbyManager*>)({parm#1}))) std::_Mem_fn_base<void (DobbyManager::*)(), true>::operator()<DobbyManager*>(DobbyManager*&&) const (functional:604)
==14000==    by 0x547360: void std::_Bind_simple<std::_Mem_fn<void (DobbyManager::*)()> (DobbyManager*)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) (functional:1391)
==14000==  Address 0xe341040 is 10,485,760 bytes inside a block of size 10,485,761 alloc'd
==14000==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14000==    by 0xC038B36: AnonymousFile::copyContentTo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (AnonymousFile.cpp:139)
==14000==    by 0xC036D6C: Minidump::postHalt() (Minidump.cpp:130)
==14000==    by 0x64B7AC: DobbyRdkPluginManager::executeHook(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, IDobbyRdkPlugin::HintFlags) const (DobbyRdkPluginManager.cpp:611)
==14000==    by 0x64BE6E: DobbyRdkPluginManager::runPlugins(IDobbyRdkPlugin::HintFlags const&) const (DobbyRdkPluginManager.cpp:717)
==14000==    by 0x534271: DobbyManager::onPostHaltHook(std::unique_ptr<DobbyContainer, std::default_delete<DobbyContainer> > const&) (DobbyManager.cpp:1830)
==14000==    by 0x534E99: DobbyManager::onChildExit() (DobbyManager.cpp:2142)
==14000==    by 0x535789: DobbyManager::runcMonitorThread() (DobbyManager.cpp:2338)
==14000==    by 0x54747A: void std::__invoke_impl<void, void (DobbyManager::* const&)(), DobbyManager*>(std::__invoke_memfun_deref, void (DobbyManager::* const&)(), DobbyManager*&&) (functional:227)
==14000==    by 0x5473F3: std::result_of<void (DobbyManager::* const&(DobbyManager*&&))()>::type std::__invoke<void (DobbyManager::* const&)(), DobbyManager*>(void (DobbyManager::* const&)(), DobbyManager*&&) (functional:251)
==14000==    by 0x547395: decltype (__invoke((*this)._M_pmf, (forward<DobbyManager*>)({parm#1}))) std::_Mem_fn_base<void (DobbyManager::*)(), true>::operator()<DobbyManager*>(DobbyManager*&&) const (functional:604)
==14000==    by 0x547360: void std::_Bind_simple<std::_Mem_fn<void (DobbyManager::*)()> (DobbyManager*)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) (functional:1391)
==14000==  Uninitialised value was created by a heap allocation
==14000==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14000==    by 0xC038B36: AnonymousFile::copyContentTo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (AnonymousFile.cpp:139)
==14000==    by 0xC036D6C: Minidump::postHalt() (Minidump.cpp:130)
==14000==    by 0x64B7AC: DobbyRdkPluginManager::executeHook(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, IDobbyRdkPlugin::HintFlags) const (DobbyRdkPluginManager.cpp:611)
==14000==    by 0x64BE6E: DobbyRdkPluginManager::runPlugins(IDobbyRdkPlugin::HintFlags const&) const (DobbyRdkPluginManager.cpp:717)
==14000==    by 0x534271: DobbyManager::onPostHaltHook(std::unique_ptr<DobbyContainer, std::default_delete<DobbyContainer> > const&) (DobbyManager.cpp:1830)
==14000==    by 0x534E99: DobbyManager::onChildExit() (DobbyManager.cpp:2142)
==14000==    by 0x535789: DobbyManager::runcMonitorThread() (DobbyManager.cpp:2338)
==14000==    by 0x54747A: void std::__invoke_impl<void, void (DobbyManager::* const&)(), DobbyManager*>(std::__invoke_memfun_deref, void (DobbyManager::* const&)(), DobbyManager*&&) (functional:227)
==14000==    by 0x5473F3: std::result_of<void (DobbyManager::* const&(DobbyManager*&&))()>::type std::__invoke<void (DobbyManager::* const&)(), DobbyManager*>(void (DobbyManager::* const&)(), DobbyManager*&&) (functional:251)
==14000==    by 0x547395: decltype (__invoke((*this)._M_pmf, (forward<DobbyManager*>)({parm#1}))) std::_Mem_fn_base<void (DobbyManager::*)(), true>::operator()<DobbyManager*>(DobbyManager*&&) const (functional:604)
==14000==    by 0x547360: void std::_Bind_simple<std::_Mem_fn<void (DobbyManager::*)()> (DobbyManager*)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) (functional:1391)
==14000== 
```

File: rdkPlugins/Minidump/source/README.md
Comment (TeknoVenus) [inline rdkPlugins/Minidump/source/README.md:4]: Please add a note that the app must be:

* Compiled with breakpad support
* Must be using the latest version of breakpad_wrapper with the `BREAKPAD_FD` support

File: rdkPlugins/Minidump/source/README.md
Comment (TeknoVenus) [inline rdkPlugins/Minidump/source/README.md:33]: Still mentions image file from previous implementation

**Review by TeknoVenus — CHANGES_REQUESTED** [review]
Please also sign the CLA (and/or add your `@sky.uk` email to your github account  as can't merge until CLA check had passed.

Some minor nitpicks on code, but otherwise happy. Assuming dumps are ending up in backtrace as they should with the plugin enabled, then once these nitpicks are fixed then happy to merge.

File: rdkPlugins/AppServices/CMakeLists.txt
Comment (g3t0r1x) [reply rdkPlugins/AppServices/CMakeLists.txt:30]: At first try I did as per your suggestion. Doing it this way caused all compilation actions to fail.

It turns out that every single Plugin is linking library `DobbyRdkPluginCommonLib`, which is fruther linking `DobbyPluginLauncherLib`. Lastly `DobbyDaemonLib` is linking library `DobbyPluginLauncherLib`. So my understanding is that if you would like to compile any `RdkPlugin` you have to compile `DobbyPluginLauncherLib` beforehand, but that receipe requires `DobbyPluginLib` interface which was not visible from parent receipe.

Anyway, the funny part is that compiling comcast image went all okay, without any issues, despite I didn't have all those CMake's modified. It only caused an issue in this repository with compile jobs. ¯\_(ツ)_/¯

File: rdkPlugins/Minidump/source/AnonymousFile.cpp
Comment (TeknoVenus) [inline rdkPlugins/Minidump/source/AnonymousFile.cpp:70]: `mFd` should be initialised to -1 here now you've added the `!= -1` check in `create()`

**Review by TeknoVenus — CHANGES_REQUESTED** [review]
Sorry - one last change please!

**Comment by bgray-sky** [comment]
The data field `DobbyStartState::mFiles` cannot be a `std::multimap` as the order of fd's are important.  By using a multimap, the numbers when mapped into the container won't match, suggest using `std::list<std::pair<std::string, int>>` instead.

**Review by TeknoVenus — APPROVED** [review]
LGTM

@bgray-sky Please can you approve if you're OK with final implementation, then I can merge.

**Review by bgray-sky — APPROVED** [review]

**Pull request merged**
